File: app/src/main/java/com/jakubwyc/mysecretvault/presenter/EditTreasurePresenter.java
package com.jakubwyc.mysecretvault.presenter;

import android.support.annotation.NonNull;

import com.jakubwyc.mysecretvault.model.Treasure;
import com.jakubwyc.mysecretvault.repository.Repository;
import com.jakubwyc.mysecretvault.view.EditTreasureView;

import rx.subscriptions.CompositeSubscription;

public class EditTreasurePresenter implements Presenter<EditTreasureView> {

    private EditTreasureView view;
    private String treasureText;
    private CompositeSubscription subscriptions = new CompositeSubscription();
    private Treasure editedTreasure;

    @Override
    public void attachView(EditTreasureView view) {
        this.view = view;
    }

    @Override
    public void detachView() {
        view = null;
        subscriptions.clear();
    }

    public void saveTreasure(final Void aVoid) {
        subscriptions.add(Repository.getInstance().saveTreasure(getTreasure())
                .subscribe($ -> view.goBackToVaultScreen()));
    }

    @NonNull
    private Treasure getTreasure() {
        return new Treasure(treasureText, editedTreasure != null ? editedTreasure.getDate() : System.currentTimeMillis());
    }

    public void treasureTextChanged(CharSequence newItemText) {
        treasureText = newItemText.toString();
    }

    public void showTreasure(final Treasure treasure) {
        editedTreasure = treasure;
        view.renderTreasure(treasure);
    }
}

File: app/src/main/kotlin/com/jakubwyc/mysecretvault/treasure/EditTreasurePresenter.kt
package com.jakubwyc.mysecretvault.treasure

import com.jakubwyc.mysecretvault.Presenter
import com.jakubwyc.mysecretvault.SystemContext
import com.jakubwyc.mysecretvault.model.Treasure
import rx.subscriptions.CompositeSubscription
import java.util.*

class EditTreasurePresenter(private val systemContext: SystemContext) : Presenter<EditTreasureView> {

    private var view: EditTreasureView? = null
    private var treasureText: String = ""
    private val subscriptions = CompositeSubscription()
    private var editedTreasure: Treasure? = null
    private val treasureRepository
        get() = systemContext.treasureRepository

    override fun attachView(view: EditTreasureView) {
        this.view = view
    }

    override fun detachView() {
        view = null
        subscriptions.clear()
    }

    fun saveTreasure() {
        subscriptions.add(treasureRepository.saveTreasure(treasure.copy())
                .subscribe { view?.goBackToVaultScreen() })
    }

    private val treasure: Treasure
        get() = Treasure(treasureText, editedTreasure?.date ?: Date().time)

    fun treasureTextChanged(newItemText: CharSequence) {
        treasureText = newItemText.toString()
    }

    fun showTreasure(treasure: Treasure) {
        editedTreasure = treasure
        view?.renderTreasure(treasure)
    }
}
